packages/approval-controller/src/ApprovalController.ts

```diff
@@ -13,13 +13,24 @@ const controllerName = 'ApprovalController';
 type ApprovalPromiseResolve = (value?: unknown) => void;
 type ApprovalPromiseReject = (error?: unknown) => void;
 
-type ApprovalRequestData = Record<string, Json> | null;
+export type ApprovalRequestData = Record<string, Json> | null;
 
 type ApprovalRequestState = Record<string, Json> | null;
 
 type ApprovalCallbacks = {
   resolve: ApprovalPromiseResolve;
   reject: ApprovalPromiseReject;
+  sideEffects?: SideEffects;
+};
+
+export type SideEffectParams = {
+  requestData: Record<string, Json> | null;
+  messagingSystem: SideEffectMessenger;
+};
+
+export type SideEffects = {
+  permittedHandlers: ((params: SideEffectParams) => Promise<unknown>)[];
+  failureHandlers: ((params: SideEffectParams) => Promise<unknown>)[];
```

Comment:
Could the `failureHandlers` be optional so you don't even need to provide the empty array?

Reply:
Initially we wanted to require both to be defined but I'm considering making it optional since failure cases can be handled before.